### ISSUE TITLE
Using Vue Apollo Composable does not correctly return mocked data.

### ISSUE THREAD

**Issue by pgoy-dod** [Mike-Gibson/mock-apollo-client#36]
## Description
When i'm using `mock-apollo-client` in my unit tests, the result from `useQuery` is always `result=undefined, loading=true, error=undefined`. The mock route is followed, but the incorrect data is returned.

## Environment

I'm currently using these libraries together:
 * @apollo/client: "^3.3.20"
 * @vue/apollo-composable": "^4.0.0-alpha.12"
 * graphql: "^15.5.1"
 * vue: "^3.0.0"
 * @vue/test-utils: "^2.0.0-0"
 * vue-jest: "^5.0.0-0"
 * mock-apollo-client: "^1.1.0"

I've got this component:
```html
//TheAccountPanel.vue
<template>
  <div id="account-panel">
    <span id="greeting">
      Hello,
      {{
        userData.firstName || userData.username || userData.email || "... you"
      }}!
    </span>
    <button id="log-out" @click="logOutUser">Log Out</button>
  </div>
</template>

<script>
import { inject } from "vue";
import { useQuery, useResult } from "@vue/apollo-composable";
import { USER } from "@/graphql/user";

export default {
  name: "TheHeader",
  setup() {
    const isLoggedIn = inject("isLoggedIn");
    const { result, loading, error, onError } = useQuery(USER);
    const userData = useResult(result, "", (data) => data.user);

    onError(() => {
      isLoggedIn.value = false;
    });

    const logOutUser = () => {
      localStorage.removeItem("apollo-token");
      isLoggedIn.value = false;
    };

    return { loading, error, userData, logOutUser };
  },
};
</script>
```

And this test:
```javascript
// TheAccountPanel.spec.js
import { ref } from "vue";
import { mount } from "@vue/test-utils";
import { provideApolloClient } from "@vue/apollo-composable";
import { createMockClient } from "mock-apollo-client";
import TheAccountPanel from "@/components/TheAccountPanel/TheAccountPanel.vue";
import { USER } from "@/graphql/user";

describe("TheAccountPanel.vue", () => {
  const isLoggedIn = ref(true);
  const BASE_USER_DATA = {
    firstName: "Firstname",
    lastName: "Lastname",
    username: "username",
    email: "email@doctorondemand.com",
  };
  let wrapper;
  let testUser = Object.assign({}, BASE_USER_DATA);
  let mockClient = createMockClient();
  const userQueryHandler = jest
    .fn()
    .mockResolvedValue({ data: { user: testUser } });
  mockClient.setRequestHandler(USER, userQueryHandler);
  provideApolloClient(mockClient);

  const createComponent = () => {
    wrapper = mount(TheAccountPanel, {
      global: {
        provide: {
          isLoggedIn: isLoggedIn,
        },
      },
    });
  };

  afterEach(() => {
    wrapper.unmount();
    isLoggedIn.value = true;
    testUser = Object.assign({}, BASE_USER_DATA);
  });

  it("uses the user's first name, if available", () => {
    createComponent();

    expect(userQueryHandler).toBeCalledTimes(1);
    expect(wrapper.find("#greeting").text()).toContain(testUser.firstName);
  });
```

The last `expect` line fails because the component always says "Hello... you!" instead of "Hello Firstname!", since the `result` was undefined.

## Investigation

I followed all the avenues i knew of to try to figure out what is happening. As you can see, the first `expect` that asserts the `userQueryHandler` had been called does actually pass. 

When i put a debugger [here](https://github.com/Mike-Gibson/mock-apollo-client/blob/a4e1d3418db30c89782abcf30f6c9c7d1be46131/src/mockLink.ts#L49), i can see this:
![proper_data_here](https://user-images.githubusercontent.com/57111606/127381405-02e57aa7-6f4d-400b-8d1e-78085fd56aa8.png)

However, this is what i see when i put a debugger in the test:
![missing_data_here](https://user-images.githubusercontent.com/57111606/127382587-ba481b2d-7ec4-4190-ab5a-5ada0e74b0e4.png)

Let me know if you need any other information. This is as close as i've gotten to a passing test with the bleeding edge potpourri of toys i'm using on this project!

**Comment by Mike-Gibson** [comment]
Thanks for the very detailed report!

I haven't tried to reproduce it yet - I won't be able to try and do that until sometime next week unfortunately. 

In the meantime, the fact that the handler is called but the Vue component doesn't show the correct response (and still thinks it's loading) makes me wonder whether the test assertion is running before the Vue component has had time to rerender with the response returned from the handler. It's something I've seen with react components, and there are different approaches to deal with it. 

I'm not too sure if the same approach will work in Vue as it does in React. A quick and dirty way to test to see if it is indeed the issue, would be to sleep for a couple of seconds before asserting the text in the Vue component, so

```js
expect(userQueryHandler).toBeCalledTimes(1);

await new Promise(r => setTimeout(r, 2000));

expect(wrapper.find("#greeting").text()).toContain(testUser.firstName)
```
I'm assuming Vue will rerender in this scenario, I might be wrong though as I haven't used Vue in a while.

Please give it a go and let me know - if it doesn't work I'll have to try and reproduce it to investigate further.

**Comment by pgoy-dod** [comment]
... Well i'll be. That did fix it! Since i was just sitting there in the debugger for so long and that didn't fix it, i had ruled out that possibility.

Is there a more... elegant way to make this test work, or will i always have to have a short pause somewhere?

**Comment by Mike-Gibson** [comment]
Glad to hear it works!

You can probably also get away with a timeout of 0 (`await new Promise(r => setTimeout(r, 0));`) as it will still cause the tets to wait for the events in the microtask queue to execute (which will probably contain the query handler result and re-render) which means your tests won't be slowed down by a load of 1 or 2 second timeouts everywhere.
In fact, this is what the official react-apollo docs [recommend](https://www.apollographql.com/docs/react/development-testing/testing/#testing-the-success-state). You could create a helper method which describes what it's doing to make the test a bit more readable, like
```js
expect(userQueryHandler).toBeCalledTimes(1);
await ensureApolloQueryResultIsRendered();
expect(wrapper.find("#greeting").text()).toContain(testUser.firstName)
```

An alternative is to use something like [wait-for-expect](https://www.npmjs.com/package/wait-for-expect) which is mentioned in the react-apollo docs above. I personally don't like the idea of using this for unit tests though, it feels like it's over the top for what's needed in unit tests.

A pattern I've used in the past is to use a custom promise class which can be used to resolve/reject the query responses and also includes the behaviour to wait for the microtask queue events, so the test code is more succint. I also tend to want to test all 3 states for components which have initial graph QL queries - loading, success and error, and so I have something like this (sorry - untested code (in TypeScript), as I don't have access to the actual implementation right now)

```typescript

class QueryPromise<T> implements Promise<T> {
    private _inner: Promise<T>;
    private _resolve: (value: T | PromiseLike<T>) => void;
    private _reject: (reason?: any) => void;

    constructor() {
        this._inner = new Promise<T>((resolve, reject) => {
            this._resolve = resolve;
            this._reject = reject;
        });
    }

    get then() {
        return this._inner.then;
    }

    get catch() {
        return this._inner.catch;
    }

    async resolve(result: T): Promise<void> {
        this._resolve(result);

        await new Promise(r => setTimeout(r, 0));
    }

    async reject(reason?: any): Promise<void> {
        this._reject(reason);

        await new Promise(r => setTimeout(r, 0));
    }
}

...

// Then in the test

describe('Some component', () => {
  let queryPromise;
  beforeEach(() => {
    queryPromise = new QueryPromise();
    queryHandler = jest
    .fn()
    .mockReturnValue(queryPromise);
    // Set up component and mock-apollo-client handler
  });

  it('renders loading text when query is loading', () => {
    // Test initial component state, and query handler has been called
  });

  describe('when query succeeds', () => {
    beforeEach(async () => {
      // The await here is subtle, but important!
      await queryPromise.resolve({ data: { name: 'bob' } });
    });

    it('renders the person's name', () => {
      // Check rendered output 
    });

    // Test actions, like clicking buttons etc
  });

// And then a similar describe block for when query fails, using `await queryPromise.reject(...)`
```

I hope that helps.
I'd be interested to know what you end up going with (especially if you find an alternative way!)

**Comment by pgoy-dod** [comment]
This is all excellent information, thank you so much for it.

After your message yesterday, i modified my tests to just wait a moment, and ran into a different problem. I was testing what the component did based on what data it got back from the query (in addition to the test above, there was one for if the user object didn't have a `firstName`, and one for if it didn't have a `firstName` or a `username`). I was having a h*ck of a time figuring out how to make the response vary based on the data i wanted to return.

I tried:
* Storing the `testData` object and modifying it per test (this was my first attempt).
  * This seems like it doesn't work because the data is cached, so no new queries are made.
* Modifying the route again with new data in each test.
  * This one didn't work because the mocked client complains about already having a route set.
* Clearing the cache between each test, and doing each of the above.
  * Neither worked! Not sure why the first one didn't work, but i have a feeling that even though i was using `Object.assign(testData, ...)`, the data is copied somewhere so it loses the link.
  
The approach I ended up taking is only using `setRequestHandler` for mutations and just directly writing the data the test needs to the `mockClient` cache, like so:

```javascript
    testUser.firstName = "";
    mockClient.writeQuery({ query: USER, data: { user: testUser } });
```

For the mutation, you're absolutely right, a timeout of `0` works. Haha. 

Thanks again for all your help! All my unit tests are passing. Great work on this mock client, it really saved a lot of headaches for me!
